File: url_detection/url_features.py
import re
from urllib.parse import urlparse

PHISHING_KEYWORDS = [
    "login","verify","update","secure","banking","account","confirm","paypal",
    "ebay","amazon","apple","microsoft","support","password","signin","wallet",
    "free","lucky","winner","claim","urgent","suspended","alert","netflix",
    "refund","prize","reward","limited","expire","blocked","kyc","otp"
]
SUSPICIOUS_TLDS = [".ru",".cn",".tk",".ml",".ga",".cf",".gq",".xyz",".top",".pw",".cc"]

def extract_url_features(url: str) -> dict:
    parsed = urlparse(url)
    return {
        "url_length":           len(url),
        "has_https":            1 if parsed.scheme=="https" else 0,
        "has_at_symbol":        1 if "@" in url else 0,
        "has_ip":               1 if re.search(r'\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}', url) else 0,
        "num_dots":             url.count("."),
        "num_hyphens":          url.count("-"),
        "num_subdomains":       max(len(parsed.netloc.split("."))-2, 0),
        "has_phishing_keyword": 1 if any(k in url.lower() for k in PHISHING_KEYWORDS) else 0,
        "has_suspicious_tld":   1 if any(url.lower().endswith(t) or (t+"/") in url.lower() for t in SUSPICIOUS_TLDS) else 0,
        "has_double_slash":     1 if url.count("//")>1 else 0,
        "path_length":          len(parsed.path),
        "has_query_params":     1 if parsed.query else 0,
    }

def score_url(url: str):
    f = extract_url_features(url)
    issues, score = [], 0
    if f["url_length"] > 75:      score+=15; issues.append("Unusually long URL")
    if not f["has_https"]:         score+=20; issues.append("No HTTPS — insecure connection")
    if f["has_at_symbol"]:         score+=25; issues.append("@ symbol (browser redirect trick)")
    if f["has_ip"]:                score+=30; issues.append("IP address instead of domain name")
    if f["num_dots"] > 4:          score+=15; issues.append("Excessive subdomains")
    if f["num_hyphens"] > 3:       score+=10; issues.append("Excessive hyphens in domain")
    if f["has_phishing_keyword"]:  score+=25; issues.append("Phishing keywords detected in URL")
    if f["has_suspicious_tld"]:    score+=25; issues.append("Suspicious TLD (.ru, .tk, .xyz etc.)")
    if f["has_double_slash"]:      score+=10; issues.append("Double-slash redirect trick")
    score = min(score, 100)
    return {
        "risk_level": "High" if score>=60 else "Medium" if score>=30 else "Low",
        "score": score,
        "prediction": "Phishing URL" if score>=60 else "Suspicious URL" if score>=30 else "Likely Safe",
        "issues": issues if issues else ["No major issues detected"]
    }

File: extension/url_features.js
// url_features.js
// Ported from url_features.py — rule-based URL risk scorer (runs entirely in the extension, no backend needed)

const PHISHING_KEYWORDS = [
  "login","verify","update","secure","banking","account",
  "confirm","paypal","ebay","amazon","apple","microsoft",
  "support","password","signin","wallet","free","lucky",
  "winner","claim","urgent","suspended","alert","netflix",
  "refund","prize","reward","limited","expire","blocked"
];

const SUSPICIOUS_TLDS = [
  ".ru",".cn",".tk",".ml",".ga",".cf",".gq",".xyz",
  ".top",".pw",".cc"
];

function extractURLFeatures(url) {
  let parsed;
  try { parsed = new URL(url.startsWith("http") ? url : "http://" + url); }
  catch(e) { parsed = { hostname: "", pathname: "", search: "", protocol: "" }; }

  const hostname = parsed.hostname || "";
  const path     = parsed.pathname || "";

  return {
    url_length:            url.length,
    has_https:             parsed.protocol === "https:" ? 1 : 0,
    has_at_symbol:         url.includes("@") ? 1 : 0,
    has_ip:                /\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}/.test(hostname) ? 1 : 0,
    num_dots:              url.split(".").length - 1,
    num_hyphens:           (url.match(/-/g) || []).length,
    num_subdomains:        Math.max(hostname.split(".").length - 2, 0),
    has_phishing_keyword:  PHISHING_KEYWORDS.some(k => url.toLowerCase().includes(k)) ? 1 : 0,
    has_suspicious_tld:    SUSPICIOUS_TLDS.some(t => url.toLowerCase().endsWith(t) || url.toLowerCase().includes(t + "/")) ? 1 : 0,
    has_double_slash:      (url.split("//").length - 1) > 1 ? 1 : 0,
    path_length:           path.length,
    has_query_params:      parsed.search ? 1 : 0,
  };
}

function scoreURL(url) {
  const f      = extractURLFeatures(url);
  const issues = [];
  let score    = 0;

  if (f.url_length > 75)          { score += 15; issues.push("Unusually long URL"); }
  if (!f.has_https)                { score += 20; issues.push("No HTTPS — insecure connection"); }
  if (f.has_at_symbol)             { score += 25; issues.push("@ symbol (browser redirect trick)"); }
  if (f.has_ip)                    { score += 30; issues.push("IP address instead of domain name"); }
  if (f.num_dots > 4)              { score += 15; issues.push("Excessive subdomains"); }
  if (f.num_hyphens > 3)           { score += 10; issues.push("Excessive hyphens in domain"); }
  if (f.has_phishing_keyword)      { score += 25; issues.push("Phishing keywords detected in URL"); }
  if (f.has_suspicious_tld)        { score += 25; issues.push("Suspicious TLD (.ru, .tk, .xyz, etc.)"); }
  if (f.has_double_slash)          { score += 10; issues.push("Double-slash redirect trick in path"); }

  score = Math.min(score, 100);

  const risk_level = score >= 60 ? "High" : score >= 30 ? "Medium" : "Low";
  const prediction = score >= 60 ? "Phishing URL" : score >= 30 ? "Suspicious URL" : "Likely Safe";

  return {
    risk_level,
    score,
    prediction,
    issues: issues.length ? issues : ["No major issues detected"]
  };
}
